discovery/gossiper.go

```diff
@@ -3256,3 +3212,83 @@ func (d *AuthenticatedGossiper) handleAnnSig(nMsg *networkMsg,
 	nMsg.err <- nil
 	return announcements, true
 }
+
+// validateChannelAnnouncement performs a series of checks over a given channel
+// announcement. It validates that,
+// - the specified chain hash is matched.
+// - alias SCID is not used for remote announcements.
+// - signature is valid for remote announcements.
+func (d *AuthenticatedGossiper) validateChannelAnnouncement(nMsg *networkMsg,
+	ann *lnwire.ChannelAnnouncement) error {
+
+	// We'll ignore any channel announcements that target any chain other
+	// than the set of chains we know of.
+	if !bytes.Equal(ann.ChainHash[:], d.cfg.ChainHash[:]) {
+		err := fmt.Errorf("ignoring ChannelAnnouncement from chain=%v"+
+			", gossiper on chain=%v", ann.ChainHash,
+			d.cfg.ChainHash)
+
+		key := newRejectCacheKey(
+			ann.ShortChannelID.ToUint64(),
+			sourceToPub(nMsg.source),
+		)
+		_, _ = d.recentRejects.Put(key, &cachedReject{})
+
+		nMsg.err <- err
+
+		return err
+	}
+
+	// We can exit it's a local announcement.
+	if !nMsg.isRemote {
+		return nil
+	}
+
+	// If this is a remote ChannelAnnouncement with an alias SCID, we'll
+	// reject the announcement. Since the router accepts alias SCIDs,
+	// not erroring out would be a DoS vector.
+	if d.cfg.IsAlias(ann.ShortChannelID) {
+		err := fmt.Errorf("ignoring remote alias channel=%v",
+			ann.ShortChannelID)
+
+		key := newRejectCacheKey(
+			ann.ShortChannelID.ToUint64(),
+			sourceToPub(nMsg.source),
+		)
+		_, _ = d.recentRejects.Put(key, &cachedReject{})
+
+		nMsg.err <- err
+
+		return err
+	}
+
+	// Otherwise we need to validate the signature of the channel
+	// announment.
+	if err := routing.ValidateChannelAnn(ann); err != nil {
+		err := fmt.Errorf("unable to validate announcement: "+
+			"%w", err)
+
+		key := newRejectCacheKey(
+			ann.ShortChannelID.ToUint64(),
+			sourceToPub(nMsg.source),
+		)
+		_, _ = d.recentRejects.Put(key, &cachedReject{})
+
+		nMsg.err <- err
```

Comment:
Hmm, what if we moved the err send point to live outside this function? This way we don't need to worry about ensuring we send the error here and we can do it at a single point. 

Eg: above on line 3244, it looks like we're forgetting to send the `nil` error over the channel.

Reply:
Totally! Previously didn't do it because we send `nMsg.err <- nil` when `isPremature` is true, and at line 3244 we don't want to send anything back yet as the function needs to continue. But the flow is not clear enough. Changed to return an error only and catch the `errFutureMsg` outside this method instead.